### PR TITLE
planner/core: support the optimize function for cascades planner

### PR DESCRIPTION
**Pull request by zz-jason**
### What problem does this PR solve? 

This PR is a split of https://github.com/pingcap/tidb/pull/7869, which contains:
- the `Optimize()` function of the cascades planner
- the `convert2Group()` function which puts a logical plan into the Memo
- the tiny refinement of the `Group` and `Pattern`

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8400)

### CODE REVIEW

**Comment by zz-jason** [comment]
@eurekaka PTAL

File: go.mod [hunk 1]
```diff
@@ -32,7 +32,7 @@ require (
 	github.com/pingcap/pd v2.1.0-rc.4+incompatible
 	github.com/pingcap/tidb-tools v0.0.0-20181112132202-4860a0d5de03
 	github.com/pingcap/tipb v0.0.0-20181012112600-11e33c750323
-	github.com/pkg/errors v0.8.0 // indirect
+	github.com/pkg/errors v0.8.0
```
Comment (eurekaka) [inline go.mod]: What is this change for?

Comment (zz-jason) [reply go.mod]: Because I explicitly imported "pkg/errors", already changed to "pingcap/errors"

**Comment by zz-jason** [comment]
@eurekaka @winoros PTAL

**Comment by zz-jason** [comment]
@lamxTyler @XuHuaiyu PTAL

**Comment by eurekaka** [comment]
@lamxTyler PTAL

**Pull request merged**